Docs/typescript.md

```diff
@@ -0,0 +1,81 @@
+# TypeScript type definitions for .NET APIs
+
+## Building type definitions
+At the moment, a `.d.ts` file is generated by the C# source generator at the same time it generates `[JSExport]` marshaling code. That is not appropriate for a C# source generator, and anyway it doesn't support generating type definitions for pre-built assemblies. So this functionality will move to a separate tool soon, along with MSBuild integration support.
+
+## Type projections reference
+
+### Primitive types
+| C# Type  | TypeScript Projection |
+|----------|-----------------------|
+| `bool`   | `boolean`             |
+| `sbyte`  | `number`              |
+| `byte`   | `number`              |
+| `short`  | `number`              |
+| `ushort` | `number`              |
+| `int`    | `number`              |
+| `uint`   | `number`              |
+| `long`   | `number`              |
+| `ulong`  | `number`              |
+| `nint`   | `number`              |
+| `nuint`  | `number`              |
+| `float`  | `number`              |
+| `double` | `number`              |
+| `string` | `string`              |
+
+### Object types
+
+| C# Type                     | TypeScript Projection       |
+|-----------------------------|-----------------------------|
+| `class Example`             | `class Example`             |
+| `struct Example`            | `class Example`             |
+| `interface IExample`        | `interface IExample`        |
+| `delegate A Example(B)`     | `example: (B) => A`         |
+| `enum Example { None = 0 }` | `enum Example { None = 0 }` |
```

Comment:
>enum [](http://example.com/codeflow?start=33&length=4)

enums are generally considered a bad practice in TypeScript and hey recommend not to use them in code because they are no zero abstractions like other type definitions. [https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#enums](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#enums)
I wonder if we should use 'const enum' or unions instead.

Reply:
> enums are generally considered a bad practice in TypeScript

Do you have a reference for that? I don't see anything in the TS handbook that says they are a bad practice.

I am familiar with the differences between `enum` and `const enum` in TypeScript, and in my experience both are useful in different situations.

For projecting C# enums to JS I planned to use non-const enums for two reasons:
  1. Const enums simply do not exist in plain JavaScript. JS code would have to use the numeric constant, which wouldn't be readable or maintainable. Or _maybe_ they could string names but that would have a perf cost and seems unnatural when C# enums are numeric.
  2. It will be useful for JS code to also be able to get the string name of an enum value when needed (e.g. for errors and diagnostics), and that can be done using the reverse mapping provided by TS style enums.

Reply:
The link that I posed above points to text that says: "Because of this, it’s a feature which you should know exists, but maybe hold off on using unless you are sure".
In Office code developers started wit using enums, but had to move to unions. I wonder if we must offer an alternative mapping for enums to developers who want to avoid the TS enum overhead.
If "TS enum" require runtime, then do we generate .js file along with the .d.ts for C# code?

Reply:
> maybe hold off on using unless you are sure

I am sure that I want plain JavaScript code to be able to reference C# enums by name.

> If "TS enum" require runtime, then do we generate .js file along with the .d.ts for C# code?

It's not needed because the runtime enum object is exported by the C# module:

```JavaScript
const assembly = require('node-api-dotnet')
    .loadAssembly('path/to/ExampleAssembly.dll');
const ExampleEnum = assembly.ExampleEnum;
const ExampleClass = assembly.ExampleClass;
ExampleClass.MethodThatTakesAnEnumParam(ExampleEnum.One);
```

Reply:
Note the TS handbook describes [problems with const enums](https://www.typescriptlang.org/docs/handbook/enums.html#const-enum-pitfalls) also.

In general const enums are great for internal use in a large project that is all TypeScript only. But for publishing enum values to external consumers, especially when they are JS, non-const enums work better.

Reply:
I added clarification in the doc that they are non-const enums.

I'm going to go ahead and check in this way, since this is documenting the way it is already implemented. We'll update the doc if we ever decide to change how C# enums are handled.